python/GafferUI/PlugValueWidget.py

```diff
@@ -443,38 +499,132 @@ def _popupMenuDefinition( self ) :
+		# support calling them if they exist.
+
+		updateMethod = getattr( self, "_updateFromPlugs", None )
+		if updateMethod is None :
+			updateMethod = getattr( self, "_updateFromPlug", None )
+
+		if updateMethod is not None :
+			## \todo Emit DeprecationWarning. It doesn't make sense to do
+			# this until we can reasonably expect Gaffer `1.1.x.x` to be
+			# out of use, as maintaining a subclass to use both the old
+			# and the new API would be incredibly painful.
+			updateMethod()
+
+	@GafferUI.LazyMethod()
+	def __callUpdateFromValues( self ) :
+
+		if self.__class__._updateFromValues is PlugValueWidget._updateFromValues :
+			# No override for `_updateFromValues()`, so no point doing all the work
+			# of calling it. Assume the subclass must be implementing the legacy
+			# `_updateFromPlug()` method instead.
+			return
+
+		with self.getContext() :
+			if any( isinstance( p, Gaffer.ValuePlug ) and Gaffer.PlugAlgo.dependsOnCompute( p ) for p in self.getPlugs() ) :
+				# Getting the values will trigger a compute, which could be
+				# arbitrarily slow. So we do it in the background to avoid locking
+				# the UI.
+				self.__updateFromValuesInBackground( self.getPlugs() )
+			else :
+				# No compute involved, so we don't expect to get any exceptions.
+				self._updateFromValues( self._valuesForUpdate( self.getPlugs() ), None )
+
+	@GafferUI.BackgroundMethod()
+	def __updateFromValuesInBackground( self, plugs ) :
+
+		return self._valuesForUpdate( plugs )
+
+	@__updateFromValuesInBackground.preCall
+	def __updateFromValuesInBackgroundPreCall( self ) :
+
+		# Many derived classes have a special "---" state when they have no
+		# plugs/values. Use this as a visual hint that a background computation
+		# is taking place.
+		self._updateFromValues( [], None )
+
+	@__updateFromValuesInBackground.plug
+	def __updateFromValuesInBackgroundPlug( self ) :
+
+		# Provide an appropriate plug for BackgroundMethod to use for
+		# cancellation.
+		return next( iter( self.getPlugs() ) )
+
+	@__updateFromValuesInBackground.postCall
+	def __updateFromValuesInBackgroundPostCall( self, backgroundResult ) :
+
+		if isinstance( backgroundResult, IECore.Cancelled ) :
+			# Cancellation. This could be due to any of the following :
+			#
+			# - This widget being hidden.
+			# - A graph edit that will affect the result and will have triggered
+			#   a call to `__callUpdateFromValues()`.
+			# - A graph edit that won't trigger a call to
+			#   `__callUpdateFromValues()`.
+			#
+			# LazyMethod takes care of all this for us. If we're hidden, it
+			# waits till we're visible. If `__callUpdateFromValues()` has
+			# already been called, our call will just replace the pending call.
+			self.__callUpdateFromValues()
+		elif isinstance( backgroundResult, Exception ) :
+			# Computation error.
+			self._updateFromValues( [], backgroundResult )
+		else :
+			# Success.
+			self._updateFromValues( backgroundResult, None )
```

Comment:
I probably ought to already understand this, but does _updateFromValues run on the UI thread?  What ensures this happens?  Does being a postCall do that?